### PR TITLE
Switch Widget Port

### PR DESCRIPTION
**Pull request by giannissc**

### CODE REVIEW

File: src/view/switch.rs
Comment (Philipp-M) [inline src/view/switch.rs:26]: I think this should probably have a callback parameter as well, for when the value is changed, similar as in the button example, but since there's a payload (`bool`), it should probably be an `impl Fn(&mut T, bool) -> A` or something like that.

So that it looks as something like this:

```rust
fn my_view(state: &mut AppState) -> impl View<AppState> {
    switch(state.my_switch_value, |state, value| { state.my_switch_value = value; })
}
```

But it would be interesting to "sugar" this callback boilerplate away somehow,
as I think something like the callback above is probably the main use-case, though I think often something additionally should happen when the value is changed as well.

For the simple case above that can probably be done via a macro? But I guess something like that should probably be done on top of this PR in a different PR...
The "UX design" of views is still a relatively big question mark so far (needs more experimentation and discussion to find the most idiomatic style IMO)
(I think SwiftUI is probably the best source of inspiration for most of this stuff)

File: src/view/switch.rs
Comment (Philipp-M) [inline src/view/switch.rs:38]: I think `bool` should actually be a generic parameter `T`, otherwise it's necessary to "`Adapt`" this everywhere, where it's used, unless the central app state is really just a `bool` :)

File: src/view/switch.rs
Comment (Philipp-M) [inline src/view/switch.rs:72]: See above, I think this should invocate the callback instead of directly acting on the `app_state` (and then return a `MessageResult::Action`).

**Review by Philipp-M — COMMENTED** [review]
Left a few comments.

I think something like an additional example showing this in action would be awesome (I think we should generally start with adding more examples, to get a feeling of the UX etc. and examples are good for testing as well)

File: src/view/switch.rs
Comment (giannissc) [reply src/view/switch.rs:72]: I figured that the callback was necessary for the button as you don't know what the intention for it is. The function of the toggle switch though is much simpler, it just needs to reverse the previous state. In there a situation that a callback would be useful?

Comment (giannissc) [reply src/view/switch.rs:72]: Question: When you add the callback to the `MessageResult` does it get executed in `message()` and the resolved value is passed up the tree using the action or do you pass the callback. If the callback is the payload of the action who executes the callback?

File: src/view/switch.rs
Comment (giannissc) [reply src/view/switch.rs:38]: Is the purpose of the callback function to be able to access specific fields from the `AppData`?

Comment (Philipp-M) [reply src/view/switch.rs:38]: Yes kind of.
It's mostly for mutating the actual app state, without restricting this to only accept `bool` as `AppState`. I think it's a slightly nicer form than to wrap it in an `Adapt` node, and "hiding" the mutation inside the `View` impl.

TMU the actual logic of mutating the `AppState` should reside in the `app_logic` function, so that it's obvious what happens. It's also a little bit more flexible, as you can "react" to the event directly in the callback, instead of "guessing" what has changed (in the next update). Not that it would make any realworld sense, but you could do something like this:

```rust
fn my_view(state: &mut AppState) -> impl View<AppState> {
    switch(state.my_switch_value, |state, value| {
        state.count += 1;
        if state.count > 3 {
            state.my_switch_value = value;
        }
    })
}
```

File: src/view/switch.rs
Comment (Philipp-M) [reply src/view/switch.rs:72]: The resolved value will just be given in the callback which then actually modifies the app state (the button example could be seen as reference here).

"Optionally" the callback can return `A` (for Action). But I feel that this will likely be used rarely (i.e. `A` will be `()`) in the actual `app_logic` (IMHO more cognitive overhead than just a callback that modifies the app state).

It's executed by passing messages currently in the widget context(s) in e.g. `Widget::event` or `Widget::update`, that in turn propagate to the view via `View::message()` using the id path that then can mutate the app state via the callback.

There are other use cases though for a `MessageResult::Action`, one being async mutations outside of the immediate `app_logic` function (and something I'd like to continue to investigate in more detail, as I feel this will be quite relevant in bigger applications).

If you're interested, I have experimented a little bit here: https://xi.zulipchat.com/#narrow/stream/354396-xilem/topic/trui. Though I would approach this currently using the `MessageResult` directly (elm like enum, to be more composable, similar as in the todomvc example).

**Comment by giannissc** [comment]
This was a good exercise! I think I understand a bit more the Xilem model. Thank for the comments they were really helpful :)

It's ready for another review! @Philipp-M

File: src/view/switch.rs
Comment (Philipp-M) [inline src/view/switch.rs:81]: I would just do something like this (like in the button view):

```suggestion
        MessageResult::Action((self.callback)(app_state, !self.is_on))
```

The main reason why I think, modifying the state directly in the callback and returning a potential `A` (instead of `&mut bool`), is that it supports `MessageResult::Action` (checkout the `main_view` function in the todomvc example in xilem_html, specifically the use of `Adapt` there).
`MessageResult::Nop` doesn't seem to be right here for me (There's happening something, i.e. the switch has been turned).
Also it would support (not that it's particularly useful) to inhibit/change the default behavior.

But the idea you're introducing here with selecting/returning the mutable reference of the boolean state inside the callback is certainly interesting though.

**Review by Philipp-M — CHANGES_REQUESTED** [review]
Just tried the example, looks nice :)

**Comment by giannissc** [comment]
Just rebased and force pushed with the new changes @Philipp-M

File: src/widget/switch.rs
Comment (Philipp-M) [inline src/widget/switch.rs:34]: (Slightly nit picky) I would name this something like `is_dragging` as I think it's more appropriate...

File: src/widget/switch.rs
Comment (Philipp-M) [inline src/widget/switch.rs:78]: (Also super nit-picky, because it shouldn't really change the behavior, as a `MouseMove` event should happen before)
```suggestion
                } else if cx.is_active() && cx.is_hot() {
```

**Review by Philipp-M — COMMENTED** [review]
Sorry for the delay, haven't found the time till now...

Some nit-picks, not really that important.

I think this is good to merge when it's rebased (for the new CI workflows) and the CI is fixed (I think you can just disable clippy for that line), thanks!

I found a bug orthogonal to this PR while testing this, maybe I'll look into this (dragging the knob or pressing a button in the `h_stack` and then moving up to the top button and releasing the cursor)

**Comment by Philipp-M** [comment]
Is this supposed to be closed?

**Comment by giannissc** [comment]
I know what happened here.. It was before I started using git extensively and I had all these into main instead of it's own branch so when synchronized my main to master I discarded my commits for the switch widget..

**Comment by jplatte** [comment]
They are still recoverable though, GitHub keeps everything that's pushed to a PR: https://github.com/linebender/xilem/commit/6275db82c73f14dce5284d7bdfb1507cdbb747f8 (parent commit are also available, and clickable in the web UI)

**Comment by giannissc** [comment]
That's ok it was easy enough to start fresh. But I will keep this in mind next time this happens. Thanks!

**Comment by Philipp-M** [comment]
I'm not sure though, whether it's possible to change the source branch (see https://stackoverflow.com/questions/42381557/changing-source-branch-for-pr-on-github)

So either you'll cherry-pick all the relevant commits onto your main branch and reopen this PR (after merging, you can of course sync with upstream main again), or you have to create a new PR with a different source branch.

File: src/widget/switch.rs
Comment (giannissc) [reply src/widget/switch.rs]: I am replying here even though I will be creating a new PR. The is_hot check I think is wrong here and the only reason it doesn't affect the behaviour is because no actual message is being send here. Had there been any message being sent to the view it would not have if the user moved past the boundaries of the widget. Good UX would be to ensure things function correctly even in light of errors from the user

**Comment by jplatte** [comment]
It's not possible to reopen a PR after the branch that was used to create it is force-pushed while the PR is closed.

File: src/widget/switch.rs
Comment (Philipp-M) [reply src/widget/switch.rs]: Hmm I guess it really depends. A button is not clicked when the user releases it outside the button area. On touchscreens I agree with you though, or generally when the switch is dragged, but this should never be the case here, since `is_dragging` will be set before, so it doesn't really matter...

**Comment by Philipp-M** [comment]
Oh really, interesting, thanks for letting me know.

**Pull request closed without merging**
